### PR TITLE
Confirm NTP ad events

### PR DESCRIPTION
**Pull request by tmancey**
Resolves 

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [x] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [ ] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- Confirm "Estimated pending rewards" reflects the correct value after viewing a new tab page ad
- Confirm "Ad notifications received this month" reflects the correct value after viewing a new tab page ad
- Confirm ad notifications are shown in the 7-days ad history
- Confirm the following confirmations for new tab page ads are triggered:
    - Viewed (maximum 4 per hour and 20 per day)
    - Clicked
    - Landed
    - Flagged (Marked as inappropriate on 7-days ad history)
    - Upvoted (Thumbs up on 7-days ad history)
    - Downvoted (Thumbs down on 7-days ad history)
    - Conversion
- Confirm clicking a new tab page ad opens a new tab for the target url
- Confirm new tab page sponsored images appear if ads is disabled
- Confirm users are not rewarded for viewing new tab page ads if ads is disabled
- Confirm new tab page ads appear after upgrading from an older version of the browser

- Confirm "Estimated pending rewards" reflects the correct value after viewing an ad notification
- Confirm "Ad notifications received this month" reflects the correct value after viewing an ad notification
- Confirm new tab page ads are shown in the 7-days ad history
- Confirm the following confirmations for ad notifications are triggered:
    - Viewed
    - Clicked
    - Dismissed
    - Landed
    - Flagged (Marked as inappropriate on 7-days ad history)
    - Upvoted (Thumbs up on 7-days ad history)
    - Downvoted (Thumbs down on 7-days ad history)
    - Conversion
- Confirm clicking an ad notification opens a new tab for the target url
- Confirm a maximum 20 ad notifications are shown per day
- Confirm a maximum 5 ad notifications are shown per hour
- Confirm ad notifications appear after upgrading from an older version of the browser

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/brave-ios/Ads/BATBraveAds.h [hunk 1]
```diff
@@ -127,10 +132,15 @@ NS_SWIFT_NAME(BraveAds)
...
 
-/// Report that a notification event type was triggered for a given id
-- (void)reportAdNotificationEvent:(NSString *)notificationUuid
+/// Report that an ad notification event type was triggered for a given id
+- (void)reportAdNotificationEvent:(NSString *)uuid
                         eventType:(BATAdNotificationEventType)eventType;
 
+/// Report that a new tab page ad event type was triggered for a given id
+- (void)reportNewTabPageAdEvent:(NSString *)uuid
+           creative_instance_id:(NSString *)creative_instance_id
```
Comment (kylehickinson) [inline vendor/brave-ios/Ads/BATBraveAds.h:141]: Small Obj-C nit: `creative_instance_id` → `creativeInstanceId`

(+ the implementation in BATBraveAds.mm)

File: vendor/brave-ios/Ads/BATBraveAds.mm [hunk 2]
```diff
@@ -474,13 +474,21 @@ - (void)reportTabClosedWithTabId:(NSInteger)tabId
...
   if (![self isAdsServiceRunning]) { return; }
-  ads->OnAdNotificationEvent(notificationUuid.UTF8String,
+  ads->OnAdNotificationEvent(uuid.UTF8String,
                              static_cast<ads::AdNotificationEventType>(eventType));
 }
 
+- (void)reportNewTabPageAdEvent:(NSString *)wallpaper_id creative_instance_id:(NSString *)creative_instance_id eventType:(BATNewTabPageAdEventType)eventType
+{
+  if (![self isAdsServiceRunning]) { return; }
+  ads->OnNewTabPageAdEvent(uuid.UTF8String,
```
Comment (kylehickinson) [inline vendor/brave-ios/Ads/BATBraveAds.mm:487]: Won't compile here, `uuid` doesn't exist (maybe just an accidental rename missed from `wallpaper_id`)

**Review by kylehickinson — CHANGES_REQUESTED** [review]
Are there further changes needed in `brave-ios`?

File: vendor/brave-ios/Ads/BATBraveAds.h (hunk 1 above)
Comment (tmancey) [reply vendor/brave-ios/Ads/BATBraveAds.h:141]: Fixed

File: vendor/brave-ios/Ads/BATBraveAds.mm (hunk 2 above)
Comment (tmancey) [reply vendor/brave-ios/Ads/BATBraveAds.mm:487]: Fixed

File: vendor/brave-ios/Ads/BATBraveAds.h [hunk 3]
```diff
@@ -127,10 +132,15 @@ NS_SWIFT_NAME(BraveAds)
...
 - (void)reportTabClosedWithTabId:(NSInteger)tabId NS_SWIFT_NAME(reportTabClosed(tabId:));
 
-/// Report that a notification event type was triggered for a given id
-- (void)reportAdNotificationEvent:(NSString *)notificationUuid
+/// Report that an ad notification event type was triggered for a given id
+- (void)reportAdNotificationEvent:(NSString *)uuid
                         eventType:(BATAdNotificationEventType)eventType;
 
+/// Report that a new tab page ad event type was triggered for a given id
+- (void)reportNewTabPageAdEvent:(NSString *)uuid
```
Comment (kylehickinson) [inline vendor/brave-ios/Ads/BATBraveAds.h:140]: `uuid` → `wallpaperId` to match implementation and usage

**Review by kylehickinson — CHANGES_REQUESTED** [review]
Just 1 last small thing, everything else looks good

**Review by kylehickinson — APPROVED** [review]
iOS lgtm

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.h [hunk 4]
```diff
@@ -298,14 +312,20 @@ class AdsImpl
   void RemoveAllAdNotificationsAfterUpdate();
   #endif
 
-  const AdNotificationInfo& get_last_shown_ad_notification() const;
-  void set_last_shown_ad_notification(
-      const AdNotificationInfo& info);
+  void set_last_clicked_ad(
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.h:315]: Can we make this CamelCase?

File: vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc [hunk 5]
```diff
@@ -203,6 +224,88 @@ std::unique_ptr<BundleState> Bundle::GenerateFromCatalog(
...
+          BLOG(1, "Creative set id " << creative_set.creative_set_id
+              << " has an invalid endAt timestamp");
+        }
+
+        info.daily_cap = campaign.daily_cap;
+        info.advertiser_id = campaign.advertiser_id;
+        info.priority = campaign.priority;
+        info.ptr = campaign.ptr;
+        info.conversion =
+            creative_set.ad_conversions.size() != 0 ? true : false;
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc:273]: Nit: Does `creative_set.ad_conversions.size() != 0` return true or false? If so, we won't need the ternary.

File: vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc [hunk 6]
```diff
@@ -203,6 +224,88 @@ std::unique_ptr<BundleState> Bundle::GenerateFromCatalog(
...
+        info.geo_targets = geo_targets;
+        info.company_name = creative.payload.company_name;
+        info.alt = creative.payload.alt;
+        info.target_url = creative.payload.target_url;
+
+        // Segments
+        for (const auto& segment : creative_set.segments) {
+          auto segment_name = base::ToLowerASCII(segment.name);
+
+          std::vector<std::string> segment_name_hierarchy =
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc:285]: Nit: I believe this allocates memory each this in this loop and the parent loop. I think if we just instantiate this once in above the parent loop, we will iterate faster.

File: vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc [hunk 7]
```diff
@@ -257,6 +361,16 @@ void Bundle::OnCreativeAdNotificationsSaved(
   BLOG(3, "Successfully saved creative ad notifications state");
 }
 
+void Bundle::OnCreativeNewTabPageAdsSaved(
+    const Result result) {
+  if (result != SUCCESS) {
+    BLOG(0, "Failed to save creative new tab page ads state");
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc:367]: Might be out-of-scope, but is it possible to add retry if it's not too hard and keep a counter on the number of retries?

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/campaigns_database_table.cc [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+  command->command = query;
+
+  transaction->commands.push_back(std::move(command));
+}
+
+void Campaigns::CreateIndexV3(
+    DBTransaction* transaction) {
+  DCHECK(transaction);
+
+  CreateIndex(transaction, get_table_name(), "campaign_id");
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/database/tables/campaigns_database_table.cc:142]: Nit: I don't think we need an index since `campaign_id` is already a PRIMARY KEY as mentioned in https://stackoverflow.com/questions/3379292/is-an-index-needed-for-a-primary-key-in-sqlite

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/categories_database_table.cc [hunk 9]
```diff
@@ -64,6 +62,11 @@ void Categories::Migrate(
       break;
     }
 
+    case 3: {
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/database/tables/categories_database_table.cc:65]: Did we skip 2? And should we have `MigrateToV1` , ``MigrateToV2`, and `MigrateToV3` as chained calls?

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/new_tab_page_ads_per_hour_frequency_cap.cc [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+std::string NewTabPageAdsPerHourFrequencyCap::get_last_message() const {
+  return last_message_;
+}
+
+bool NewTabPageAdsPerHourFrequencyCap::DoesRespectCap(
+    const std::deque<uint64_t>& history) {
+  const uint64_t time_constraint = base::Time::kSecondsPerHour;
+
+  const uint64_t cap = 4;
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/new_tab_page_ads_per_hour_frequency_cap.cc:43]: Nit: Make this and `time_constaint` class variables

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table.h [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+      AdsImpl* ads);
+
+  ~CreativeNewTabPageAds() override;
+
+  void Save(
+      const CreativeNewTabPageAdList& creative_new_tab_page_ads,
+      ResultCallback callback);
+
+  void GetCreativeNewTabPageAd(
+      const std::string& creative_instance_id,
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table.h:49]: Maybe this is a big refactor that should be handled in a separate PR, but it's throwing me off that we have shared ID's without corresponding tables for `creative_instance_id`, and`creative_set_id`. Additionally, I think it makes more sense to call this `creative_id`, as `instance` is implied.

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table_unittest.cc [hunk 12]
```diff
@@ -0,0 +1,864 @@
...
+  info.total_max = 4;
+  info.category = "Technology & Computing-Software";
+  info.geo_targets = { "US" };
+  info.target_url = "https://brave.com";
+  info.title = "Test Ad 1 Title";
+  info.body = "Test Ad 1 Body";
+  info.ptr = 1.0;
+  creative_ad_notifications.push_back(info);
+
+  SaveDatabase(creative_ad_notifications);
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table_unittest.cc:390]: Double `SaveDatabase`

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table.cc [hunk 13]
```diff
@@ -0,0 +1,533 @@
...
+  if (creative_new_tab_page_ads.empty()) {
+    callback(Result::SUCCESS);
+    return;
+  }
+
+  DBTransactionPtr transaction = DBTransaction::New();
+
+  // TODO(https://github.com/brave/brave-browser/issues/3661) instead of
+  // rebuilding the catalog database each time
+  Delete(transaction.get(), get_table_name());
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table.cc:61]: Perhaps use `Drop()` instead?

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_ad_notifications_database_table.cc [hunk 14]
```diff
@@ -56,15 +58,23 @@ void CreativeAdNotifications::Save(
 
   // TODO(https://github.com/brave/brave-browser/issues/3661) instead of
   // rebuilding the catalog database each time
-  DeleteAllTables(transaction.get());
+  Delete(transaction.get(), get_table_name());
```
Comment (yachtcaptain23) [inline vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_ad_notifications_database_table.cc:61]: Perhaps use `Drop()` here too

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.h (hunk 4 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.h:315]: Discussed and agreed to resolve

File: vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc (hunk 5 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc:273]: Discussed and agreed to resolve

File: vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc (hunk 6 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc:285]: Discussed and not an issue

File: vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc (hunk 7 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/bundle/bundle.cc:367]: Discussed and we retry the next time we download a catalog

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/categories_database_table.cc (hunk 9 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/database/tables/categories_database_table.cc:65]: Discussed and agreed

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/new_tab_page_ads_per_hour_frequency_cap.cc (hunk 10 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/new_tab_page_ads_per_hour_frequency_cap.cc:43]: Fixed cap, thanks

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table.h (hunk 11 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table.h:49]: Discussed and agreed to rename `creative_ads` table to `creative_sets`

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table_unittest.cc (hunk 12 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table_unittest.cc:390]: Discussed and not an issue

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table.cc (hunk 13 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_new_tab_page_ads_database_table.cc:61]: As we drop the bundle database every time we update the catalog I agree we should remove the migration code and recreate the table. However, as discussed this should be in another issue/PR

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_ad_notifications_database_table.cc (hunk 14 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_ad_notifications_database_table.cc:61]: see last comment

**Review by yachtcaptain23 — APPROVED** [review]
Discussed changes with @tmancey . His updated commit will resolve all my questions

**Review by moritzhaller — APPROVED** [review]
verified workings according to test plan as demonstrated by @tmancey - LGTM!
